File: backend/restaurants/src/service/minio-service.ts
import {Restaurant} from '.prisma/client'
import * as Minio from 'minio'
import {Client, UploadedObjectInfo} from 'minio'
import {Readable} from 'stream'

class MinioService {
    client: Client
    bucketName: string
    embedBucketName: string

    constructor() {
        // Local configuration
        if (process.env.USE_MINIO_LOCAL) {
            if (
                process.env.MINIO_ACCESS_KEY == undefined ||
                process.env.MINIO_SECRET_KEY == undefined
            )
                throw new Error('Local MINIO credentials not found')

            this.client = new Minio.Client({
                endPoint: 'minio-restaurants',
                port: 9002,
                accessKey: process.env.MINIO_ACCESS_KEY,
                secretKey: process.env.MINIO_SECRET_KEY,
                useSSL: false,
            })
            this.bucketName = 'images'
            this.embedBucketName = 'embeddings'
        } else {
            // AWS S3 configuration
            if (
                process.env.AWS_ACCESS_KEY == undefined ||
                process.env.AWS_SECRET_KEY == undefined ||
                process.env.AWS_SESSION_TOKEN == undefined
            )
                throw new Error('AWS credentials not found')

            this.client = new Minio.Client({
                endPoint: 's3.amazonaws.com',
                accessKey: process.env.AWS_ACCESS_KEY,
                secretKey: process.env.AWS_SECRET_KEY,
                sessionToken: process.env.AWS_SESSION_TOKEN,
            })
            this.bucketName = 'voicefork-restaurants-images'
            this.embedBucketName = 'voicefork-restaurant-embeddings' //TODO change
        }
    }

    async putObject(objectName: string, object: string | Readable | Buffer) {
        return this.client.putObject(
            this.bucketName,
            objectName,
            object,
            (err: Error, objInfo: UploadedObjectInfo) => {
                if (err) {
                    console.log(err)
                }
                console.log('Object successfully stored. Object info:', objInfo)
                return objInfo
            },
        )
    }

    async getObject(objectName: string, bucketName: string): Promise<Buffer> {
        const chunks: Buffer[] = []

        const dataStream = await this.client.getObject(bucketName, objectName)

        dataStream.on('data', (chunk: Buffer) => {
            chunks.push(chunk)
        })

        return new Promise<Buffer>((resolve, reject) => {
            dataStream.on('end', () => {
                const objectData = Buffer.concat(chunks)
                console.log('Object successfully retrieved:', objectData)
                resolve(objectData)
            })

            dataStream.on('error', (error: Error) => {
                console.error('Error reading object data stream:', error)
                reject(error)
            })
        })
    }

    async getImage(imageName: string): Promise<string> {
        const image = (
            await this.getObject(imageName, this.bucketName)
        ).toString('base64')
        return `data:image/jpeg;base64,${image}`
    }

    async getEmbedding(embeddingName: string): Promise<number[]> {
        const embedding = (
            await this.getObject(embeddingName, this.embedBucketName)
        ).toString('utf-8')
        return JSON.parse(embedding)
    }
}

export default MinioService

File: backend/embeddings/services/MinioService.py
from minio import Minio
from dotenv import load_dotenv
import os
import logging

logger = logging.getLogger(__name__)

load_dotenv()

USE_MINIO_LOCAL = os.getenv("USE_MINIO_LOCAL")
AWS_ACCESS_KEY = os.getenv("AWS_ACCESS_KEY")
AWS_SECRET_KEY = os.getenv("AWS_SECRET_KEY")
AWS_SESSION_TOKEN = os.getenv("AWS_SESSION_TOKEN")

MINIO_ACCESS_KEY = os.getenv("MINIO_ACCESS_KEY")
MINIO_SECRET_KEY = os.getenv("MINIO_SECRET_KEY")
MINIO_IP = os.getenv("MINIO_IP")


class MinioService:
    def __init__(self):
        if USE_MINIO_LOCAL == "true":
            self.client = Minio(
                "minio-restaurants:9002",
                access_key=MINIO_ACCESS_KEY,
                secret_key=MINIO_SECRET_KEY,
                secure=False
            )
            self.BUCKET_NAME = "embeddings"
            self.NUM_OBJECTS = 222603
        else:
            self.client = Minio(
                "s3.amazonaws.com",
                access_key=AWS_ACCESS_KEY,
                secret_key=AWS_SECRET_KEY,
                session_token=AWS_SESSION_TOKEN)
            self.BUCKET_NAME = "voicefork-embeddings-bucket"
            self.NUM_OBJECTS = 0

    def get_embedding(self, embedding_name: str):
        response = self.client.get_object(self.BUCKET_NAME, embedding_name)
        data = response.read()

        response.close()
        response.release_conn()

        return data

    def count_objects(self) -> int:
        logger.info("Counting objects")
        objects = self.client.list_objects(self.BUCKET_NAME)
        self.NUM_OBJECTS = sum(1 for _ in objects)
        logger.info(
            f"Finish counting, there are {self.NUM_OBJECTS} objects in the {self.BUCKET_NAME} bucket")
        return self.NUM_OBJECTS
